File: internal/httpapi/folders.go
package httpapi

import (
	"errors"
	"net/http"

	"github.com/go-freya/freya/services/warden/internal/authz"
	"github.com/go-freya/freya/services/warden/internal/folders"
	"github.com/go-freya/freya/services/warden/internal/secrets"
)

// StoryDeps are the services behind the folder and secret routes.
type StoryDeps struct {
	Folders *folders.Service
	Secrets *secrets.Service
	Authz   *authz.Authz
}

// subjects derives the authz subjects from the verified caller.
func subjects(r *http.Request) (authz.Subjects, error) {
	id, err := Caller(r)
	if err != nil {
		return authz.Subjects{}, err
	}
	return authz.SubjectsOf(id), nil
}

// domainError maps service errors to refusals.
func domainError(err error) error {
	switch {
	case errors.Is(err, authz.ErrForbidden):
		return ErrForbidden
	case errors.Is(err, authz.ErrNotFound), errors.Is(err, secrets.ErrNoTOTP):
		return ErrNotFound
	case errors.Is(err, authz.ErrInput), errors.Is(err, folders.ErrInvalidName), errors.Is(err, secrets.ErrInvalid), errors.Is(err, secrets.ErrInvalidTOTP):
		return ErrValidation
	case errors.Is(err, folders.ErrConflict), errors.Is(err, folders.ErrNotEmpty), errors.Is(err, folders.ErrCycle):
		return ErrConflict
	}
	return err
}

// RegisterFolders mounts the folder routes.
func (s *Server) RegisterFolders(d StoryDeps) {
	s.MustHandle("GET", "/api/warden/v1/folders", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var parent *string
		if p := r.URL.Query().Get("parent_id"); p != "" {
			parent = &p
		}
		out, err := d.Folders.Children(r.Context(), subj, parent)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, map[string]any{"items": out})
	})
	s.MustHandle("POST", "/api/warden/v1/folders", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			ParentID *string `json:"parent_id"`
			Name     string  `json:"name"`
		}
		if err := DecodeJSON(r, &in, 0); err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Folders.Create(r.Context(), subj, in.ParentID, in.Name)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusCreated, out)
	})
	s.MustHandle("GET", "/api/warden/v1/folders/tree", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Folders.Tree(r.Context(), subj)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, map[string]any{"items": out})
	})
	s.MustHandle("GET", "/api/warden/v1/folders/{id}", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Folders.Get(r.Context(), subj, r.PathValue("id"))
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("PUT", "/api/warden/v1/folders/{id}", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			Name string `json:"name"`
		}
		if err := DecodeJSON(r, &in, 0); err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Folders.Rename(r.Context(), subj, r.PathValue("id"), in.Name)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("POST", "/api/warden/v1/folders/{id}/move", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			ParentID *string `json:"parent_id"`
		}
		if err := DecodeJSON(r, &in, 0); err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Folders.Move(r.Context(), subj, r.PathValue("id"), in.ParentID)
		if err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("POST", "/api/warden/v1/folders/{id}/remove", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			Recursive bool `json:"recursive"`
		}
		if r.ContentLength != 0 {
			if err := DecodeJSON(r, &in, 0); err != nil {
				Fail(w, r, nil, err)
				return
			}
		}
		if err := d.Folders.Delete(r.Context(), subj, r.PathValue("id"), in.Recursive); err != nil {
			Fail(w, r, s.rt.Logger(), domainError(err))
			return
		}
		w.WriteHeader(http.StatusNoContent)
	})
}

File: ui/src/stores/folders.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '@/api/client'
import type { Folder, FolderNode } from '@/api/types'

export const useFolders = defineStore('warden-folders', () => {
  const tree = ref<FolderNode[]>([])
  const loading = ref(false)
  const error = ref('')

  async function load(): Promise<void> {
    loading.value = true
    error.value = ''
    try {
      const out = await api<{ items: FolderNode[] }>('GET', 'folders/tree')
      tree.value = out.items
    } catch (e) {
      error.value = (e as Error).message
    } finally {
      loading.value = false
    }
  }

  function find(id: string | null, nodes: FolderNode[] = tree.value): FolderNode | undefined {
    for (const n of nodes) {
      if (n.folder.id === id) return n
      const hit = find(id, n.children)
      if (hit) return hit
    }
    return undefined
  }

  async function create(parentId: string | null, name: string): Promise<Folder> {
    const f = await api<Folder>('POST', 'folders', { parent_id: parentId, name })
    await load()
    return f
  }

  async function rename(id: string, name: string): Promise<Folder> {
    const f = await api<Folder>('PUT', 'folders/' + id, { name })
    await load()
    return f
  }

  async function move(id: string, parentId: string | null): Promise<Folder> {
    const f = await api<Folder>('POST', 'folders/' + id + '/move', { parent_id: parentId })
    await load()
    return f
  }

  async function remove(id: string, recursive: boolean): Promise<void> {
    await api('POST', 'folders/' + id + '/remove', { recursive })
    await load()
  }

  /** Flattens the tree depth-first for pickers. */
  function flat(nodes: FolderNode[] = tree.value, depth = 0): Array<{ folder: Folder; depth: number }> {
    const out: Array<{ folder: Folder; depth: number }> = []
    for (const n of nodes) {
      out.push({ folder: n.folder, depth })
      out.push(...flat(n.children, depth + 1))
    }
    return out
  }

  return { tree, loading, error, load, find, create, rename, move, remove, flat }
})
